Tzkt.Api/Websocket/Processors/CyclesProcessor.cs

```diff
@@ -0,0 +1,151 @@
+﻿using System;
+using System.Linq;
+using System.Collections.Generic;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.AspNetCore.SignalR;
+using Microsoft.Extensions.Logging;
+
+using Tzkt.Api.Repositories;
+using Tzkt.Api.Services.Cache;
+using Tzkt.Api.Models;
+
+namespace Tzkt.Api.Websocket.Processors
+{
+    public class CycleProcessor<T> : IHubProcessor where T : Hub
+    {
+        #region static
+        const string CycleGroup = "cycle";
+        const string CycleChannel = "cycle";
+        static readonly SemaphoreSlim Sema = new(1, 1);
+        static readonly Dictionary<int, HashSet<string>> DelaySubs = new();
+        static readonly Dictionary<string, Symbols> QuoteSubs = new();
+        #endregion
+
+        readonly StateCache StateCache;
+        readonly CyclesRepository CyclesRepo;
+        readonly IHubContext<T> Context;
+        readonly ILogger Logger;
+
+        public CycleProcessor(StateCache cache, CyclesRepository repo, IHubContext<T> hubContext, ILogger<CycleProcessor<T>> logger)
+        {
+            StateCache = cache;
+            CyclesRepo = repo;
+            Context = hubContext;
+            Logger = logger;
+        }
+
+        public async Task OnStateChanged()
+        {
+            var sendings = new List<Task>(2);
+            try
+            {
+                await Sema.WaitAsync();
+
+                var cycleData = await CyclesRepo.Get(StateCache.Current.Cycle, Symbols.None);
+
+                // we notify only group of clients with matching delay
+                if (DelaySubs.TryGetValue(StateCache.Current.Level - cycleData.FirstLevel, out var connections))
+                {
+                    foreach (var connectionId in connections)
+                    {
+                        var quote = QuoteSubs[connectionId];
+                        var cycleDataWithQuotes = quote == Symbols.None ? cycleData : await CyclesRepo.Get(StateCache.Current.Cycle, quote);
```

Comment:
If there are 1000 connections with configured quotes, the processor will do 1001 requests to the database every block. That's quite dangerous, as the number of connections may be even greater.

Reply:
Added QuoteCache.